### ISSUE TITLE
Implement a 'keystore' were a user to can add certificates to authorise clients

### ISSUE THREAD

**Issue by rcrichton** [jembi/openhim-core-js#299]
We need to add a keystore to store certificates that are used to authorise clients. Here is a use case that I wrote (quickly, so mind the grammar...)

> I'm really thinking about this in the case of the IHE connectathon. I that case they give each client a cert. Each those certs are signed by their own connectatnon Certificate Authority (CA). So as the OpenHIM, I need to be able to say I accept all client that are signed by the CA (I won't know about each client's cert). We can do this by telling node we accept all client's that have certificates that are signed by the CA, this is done by importing the CA cert. Then for us to determine who the client is, we look at the certificate that they present, and match them on the CN of the cert and the domain of the client in our datastore.

**Comment by rcrichton** [comment]
So for this issue here is what I think the keystore should do:

* Allow the admin user to upload a custom key and cert for the OpenHIM
* Allow the admin user to upload individual trusted certs (for clients or hosts)
* Allow the admin user to upload trusted certificate chains (for clients or hosts)
* Allow the admin user to generate a self signed cert for a OpenHIM implementation rather than having to upload one
* Allow the admin user to generate a key and certificate for a client signed by the HIM cert

**Comment by rcrichton** [comment]
The certificates could be read and/or generated using something like: https://www.npmjs.com/package/pem

Also, to add to the node root CAs we could add to the `https.globalAgent.options.ca` as they do in this project: https://github.com/coolaj86/node-ssl-root-cas This will allow us to use both self-signed trusted CAs and the well know root CAs. (in theory)

**Comment by rcrichton** [comment]
To store the certificate info we could have a singleton mongo collection as follows:

collection: `keystore` (`truststore`?)
```
{
  key: '<pem string>',
  cert: '<pem string>',
  ca: [
    '<pem string>',
    '<pem string>',
    '<pem string>'
    OR
    {
       subject: { cn: '', O: '', OU: '', etc...},
       issuer: { cn: '', O: '', OU: '', etc...},
       validity: { ... },
       cert: '<pem string>'
    },
    {
       subject: { cn: '', O: '', OU: '', etc...},
       issuer: { cn: '', O: '', OU: '', etc...},
       validity: { ... },
       cert: '<pem string>'
    }
  ]
}
```

Should the CA certs be stored with details such as subject or issuer, or should this be extracted on demand? We  could extract on demand on the server side, in the API function.

**Comment by rcrichton** [comment]
The CA certs would get added to the ca property for both inbound and outbound requests. Thus both clients are servers that we need to communicate with will be covered in a single place. We should always trust root CAs as well, however, for clients we still associate clients with channels so that only authorised inbound requests are accepted. Client are linked to a certificate by their `domain` property.

**Comment by rcrichton** [comment]
Suggested API endpoints:

`GET keystore/cert` - get the current HIM server cert
returns 200 ok with `{ subject: '', issuer: '', validity: '', cert: '<pem string>' }`

`GET keystore/ca` - gets the array of trusted ca certs
returns 200 ok with `[ { subject: '', issuer: '', validity: '', cert: '<pem string>' }, ... ]`

`GET keystore/ca/_id` - gets a ca cert by its _id
returns 200 ok with `{ subject: '', issuer: '', validity: '', cert: '<pem string>' }`

`POST keystore/key` - sets the HIM server key
data `{ key: '<pem string>' }`
returns 201 ok

`POST keystore/cert` - sets the HIM server cert
data `{ cert: '<pem string>' }`
returns 201 ok

`POST keystore/ca/cert` - adds a cert to the list of ca certs
data `{ cert: '<pem string>' }`
returns 201 ok

`POST keystore/ca` - sets a new array of ca certs
data `[ '<pem string>', '<pem string>', ... ]`
returns 201 ok

`DELETE keystore/ca/_id` - removes a ca cert by its _id
returns 200 ok

**Comment by bausmeier** [comment]
The way that I see this working is when creating a client or route, instead of having a textarea to enter the contents of a certificate pem file, rather to have a dropdown to select a certificate which has already been added to the list of trusted certificates. These could be identified by the common name of the certificate or by some other combination of the subject fields.

We could also have an option to allow all of the certificates in the list to be used for a route i.e. add them all as trust anchors along with the default ones used by node.

How would we reference the certificates from the route/client? How would this handle updating or removing certificates from the trusted list?

**Comment by rcrichton** [comment]
Thanks @bausmeier I agree this would be great. I'd suggest we link a route to the _id of a trusted certificate instead of having the raw cert data in the route. Then if a cert no longer exists we should return an error for that route stating that the trust certificate cannot be found, please update the route.

**Comment by rcrichton** [comment]
Alos, I'm going to split the generating of client certificates with the OpenHIM acting as a CA into a separate issue. It seems it could be more complex than we think from @bausmeier's comments over hipchat.

**Comment by bausmeier** [comment]
That makes sense. Or maybe we could not let you delete the cert while it is
still referenced? That would be slightly more work to implement though.

**Comment by rcrichton** [comment]
Yeah that is also an option. But, I think I'd rather the admin be able to manage the trusted certs easily rather than clear up that reference. For fear of them not bothering to remove certificates that shouldn't be trusted anymore.

**Comment by devcritter** [comment]
@rcrichton sorry for taking so long to take a look at this. All in all it looks good to me, although afraid I'm not a big enough expert in PKI in order to offer too many comments!

This may be a dumb question, but in the mongo entries do you plan on reading the subject, issuer, etc. fields from the pem? Or would the user need to specify that somehow?

In the former case I think's it definitely good to have those fields in the model, plus it adds a extra level of validation. So the API can respond saying it's an invalid key if it can't read those fields...

**Comment by rcrichton** [comment]
@devcritter I plan to read those from the pem file itself using this: https://www.npmjs.com/package/pem

Yeah, I agree, it will help to ensure the certificates that we upload are valid.

**Comment by devcritter** [comment]
Aah, ok great

**Comment by rcrichton** [comment]
Closed via #359
